### PR TITLE
UI tests

### PR DESCRIPTION
**Pull request by riccjohn**
- Adds accessibility identifiers to UI elements
- Adds UI tests

### CODE REVIEW

File: Junk4Dollars-appUITests/Junk4Dollars_appUITests.swift [hunk 1]
```diff
@@ -0,0 +1,51 @@
+//
+//  Junk4Dollars_appUITests.swift
+//  Junk4Dollars-appUITests
+//
+//  Created by John Riccardi on 10/28/19.
+//  Copyright © 2019 John Riccardi. All rights reserved.
+//
```
Comment (ijpiantanida) [inline Junk4Dollars-appUITests/Junk4Dollars_appUITests.swift]: Mind changing the settings to remove this automatic comment?

File: Junk4Dollars-appUITests/Junk4Dollars_appUITests.swift [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        let textInput = app.textFields["text_input"]
+        textInput.tap()
+        textInput.typeText("John")
+        app.buttons["hello_btn"].tap()
+        app.buttons["reset_btn"].tap()
+        app.buttons["hello_btn"].tap()
+        XCTAssert(app.staticTexts["Hello!"].exists)
+    }
+
+    func testLaunchPerformance() {
```
Comment (ijpiantanida) [inline Junk4Dollars-appUITests/Junk4Dollars_appUITests.swift:46]: What exactly does this do? Does it just log the metrics?

File: Junk4Dollars-appUITests/Junk4Dollars_appUITests.swift [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+//  Created by John Riccardi on 10/28/19.
+//  Copyright © 2019 John Riccardi. All rights reserved.
+//
+
+import XCTest
+
+class Junk4Dollars_appUITests: XCTestCase {
+    func testHelloDefault() {
+        let app = XCUIApplication()
+        app.launch()
```
Comment (ijpiantanida) [inline Junk4Dollars-appUITests/Junk4Dollars_appUITests.swift]: How can you remove this duplication from each test?

File: Junk4Dollars-appUITests/Junk4Dollars_appUITests.swift [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        let textInput = app.textFields["text_input"]
+        textInput.tap()
+        textInput.typeText("John")
+        app.buttons["hello_btn"].tap()
+        let helloJohnLabel = app.staticTexts["Hello, John!"]
+        XCTAssert(helloJohnLabel.exists)
+    }
+    
+    func testResetButton() {
```
Comment (ijpiantanida) [inline Junk4Dollars-appUITests/Junk4Dollars_appUITests.swift]: Test names should explain what the test is doing and looking for.
In this case, something like `testResetButtonClearsOutput` would be better.

Also, the reset button is doing something else that it's not being tested. What is it?

File: Junk4Dollars-appUITests/Junk4Dollars_appUITests.swift (hunk 2 above)
Comment (riccjohn) [reply Junk4Dollars-appUITests/Junk4Dollars_appUITests.swift:46]: Yeah, it opens and closes the app 6 times and logs the metrics to make sure the load times don't start increasing too much.

File: Junk4Dollars-appUITests/Junk4Dollars_appUITests.swift (hunk 4 above)
Comment (riccjohn) [reply Junk4Dollars-appUITests/Junk4Dollars_appUITests.swift]: Changed the name and added a test for `testResetButtonClearsInput`

File: Junk4Dollars-appUITests/Junk4Dollars_appUITests.swift (hunk 3 above)
Comment (riccjohn) [reply Junk4Dollars-appUITests/Junk4Dollars_appUITests.swift]: Moved the app launch to a setup function

File: Junk4Dollars-appUITests/Junk4Dollars_appUITests.swift (hunk 1 above)
Comment (riccjohn) [reply Junk4Dollars-appUITests/Junk4Dollars_appUITests.swift]: It looks like this involves copying / altering templates from the Xcode app itself. Do you think it's worth it? https://stackoverflow.com/questions/33720/change-templates-in-xcode/33743#33743

File: Junk4Dollars-appUITests/Junk4Dollars_appUITests.swift [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        textInput.tap()
+        let textToType = "John"
+        textInput.typeText(textToType)
+        app.buttons["hello_btn"].tap()
+        app.buttons["reset_btn"].tap()
+        app.buttons["hello_btn"].tap()
+        XCTAssert(app.staticTexts["Hello!"].exists)
+    }
+    
+    func testResetButtonClearsInput() {
```
Comment (ijpiantanida) [inline Junk4Dollars-appUITests/Junk4Dollars_appUITests.swift]: Looks like the test is missing the "tap reset" part

Comment (riccjohn) [reply Junk4Dollars-appUITests/Junk4Dollars_appUITests.swift]: Oops. fixed

File: Junk4Dollars-appUITests/Junk4Dollars_appUITests.swift [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        let textInput = app.textFields["text_input"]
+        textInput.tap()
+        let textToType = "John"
+        textInput.typeText(textToType)
+        app.buttons["reset_btn"].tap()
+        XCTAssertEqual(textInput.value as! String, "Name")
```
Comment (ijpiantanida) [inline Junk4Dollars-appUITests/Junk4Dollars_appUITests.swift:43]: Where is this `Name` coming from? Does't resetting clear out the input?

Comment (riccjohn) [reply Junk4Dollars-appUITests/Junk4Dollars_appUITests.swift:43]: Name is the element's placeholder text. I was trying to assert it was equal to `""`, but it returns `Name`

File: Junk4Dollars-appTests/ViewControllerTests.swift [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        
+        controller.viewDidLoad()
+        
+        XCTAssertEqual("", controller.label.text)
+    }
+    
+    func testSayHelloDefault() {
+        let controller = ViewController()
+        controller.label = UILabel()
+        controller.textInput = UITextField()
```
Comment (ijpiantanida) [inline Junk4Dollars-appTests/ViewControllerTests.swift]: This initialization is the same on every test. Remove duplication?

File: Junk4Dollars-appTests/ViewControllerTests.swift [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        controller.textInput.text = "John"
+        
+        controller.tapSayHello()
+        
+        XCTAssertEqual("Hello, John!", controller.label.text)
+    }
+    
+    func testResetButton() {
```
Comment (ijpiantanida) [inline Junk4Dollars-appTests/ViewControllerTests.swift]: Same comment as with the UI test, test names should explain the scenario + expected result, so you don't have to dig deep into the test to understand what it's doing.

**Review by ijpiantanida — APPROVED** [review]

**Pull request closed without merging**